src/core_plugins/kibana/public/visualize/editor/editor.js

```diff
@@ -316,20 +318,23 @@ function VisEditor($rootScope, $scope, $route, timefilter, AppState, $window, kb
       if (id) {
         notify.info('Saved Visualization "' + savedVis.title + '"');
         if ($scope.isAddToDashMode()) {
-          const savedVisualizationUrl =
-            kbnUrl.eval(`${kbnBaseUrl}#${VisualizeConstants.EDIT_PATH}/{{id}}`, { id: savedVis.id });
-
+          const savedVisualizationParsedUrl = new KibanaParsedUrl({
+            basePath: chrome.getBasePath(),
+            appId: kbnBaseUrl.slice('/app/'.length),
+            appPath: kbnUrl.eval(`${VisualizeConstants.EDIT_PATH}/{{id}}`, { id: savedVis.id })
+          });
           // Manually insert a new url so the back button will open the saved visualization.
-          $window.history.pushState({}, '', `${chrome.getBasePath()}${savedVisualizationUrl}`);
+          $window.history.pushState({}, '', savedVisualizationParsedUrl.getRootRelativePath());
           // Since we aren't reloading the page, only inserting a new browser history item, we need to manually update
           // the last url for this app, so directly clicking on the Visualize tab will also bring the user to the saved
           // url, not the unsaved one.
-          chrome.trackSubUrlForApp('kibana:visualize', savedVisualizationUrl);
+          chrome.trackSubUrlForApp('kibana:visualize', savedVisualizationParsedUrl);
+
+          const lastDashboardAbsoluteUrl = chrome.getNavLinkById('kibana:dashboard').lastSubUrl;
+          const dashboardParsedUrl = absoluteToParsedUrl(lastDashboardAbsoluteUrl, chrome.getBasePath());
+          const appPathWithNewQuery = `${dashboardParsedUrl.appPath}&${DashboardConstants.NEW_VISUALIZATION_ID_PARAM}={{id}}`;
```

Comment:
I thought about adding functionality to KibanaParsedUrl like addQueryParameter(name, value) but I would have to extract kbnUrl.eval out into a separate function.  I think this is worthwhile, but didn't want to get too down the rabbit hole with this PR, and left that as a follow up.